### ISSUE TITLE
ChannelHangupComplete event

### ISSUE THREAD

**Issue by mguerrieri** [danbarua/NEventSocket#60]
I would like to propose that a HangupCompleteCallback handler be added to the Channel API, moving the Dispose() call here from the HangupCallback() handler.  The reason being, in my particular application I need to log CDR data from within the app, and this data is presented during the ChannelHangupComplete event.  Unfortunately, as currently coded, the channel is disposed when the ChannelHangup event is detected which causes the channel to be disposed before the ChannelHangupComplete event can be handled.

**Comment by danbarua** [comment]
Hi,

We're not _entirely_ clear on the scenarios where this occurs, but certainly in my testing I've found that I never received the CHANNEL_HANGUP_COMPLETE event reliably, which is why I use CHANNEL_HANGUP instead.

Related #52

http://lists.freeswitch.org/pipermail/freeswitch-users/2014-January/102397.html
http://lists.freeswitch.org/pipermail/freeswitch-users/2013-June/096224.html

is it possible to use a dedicated InboundSocket to do the CDR logging, or is this intertwined with your call flow logic?

**Comment by danbarua** [comment]
Of course, I would love to be wrong on this if we can figure out what the issue is.

**Comment by mguerrieri** [comment]
Well I probably could use an inbound socket- I have one open for originating calls.  But I will do some testing and see if I can find scenarios where the COMPLETE event doesn't arrive.  Given that the FreeSWITCH docs all indicate that handling CDR via ESL should be done by handling this even, you would hope that the even is actually sent consistently.

**Comment by danbarua** [comment]
I've just tested this and it seems to work. I'll keep this issue open while we test at $dayJob.

**Comment by mguerrieri** [comment]
Hmm... in my testing it does NOT seem to work- I am not seeing the COMPLETE event on the outbound socket regardless of whether I drop the call from FS or the remote end.

**Comment by mguerrieri** [comment]
Never mind- user error. Seeing the COMPLETE event now.

**Comment by danbarua** [comment]
Would you mind sharing what the user error was? There's at least myself and one other who've had a similar experience with the COMPLETE event.

**Comment by mguerrieri** [comment]
Was working off a local copy of the source and applied your changes manually, but missed the line in the change subscribed to the COMPLETE event.  Once I added that, I received the COMPLETE event in all of my tests (few as they were…)
